### PR TITLE
SOC-5283 : Improve social search connecto

### PR DESCRIPTION
**Pull request by aboughzela**
Improvement PeopleSearchConnector and SpaceSearchConnector : 
by Default social workspace doesn't contain a configured analyser
In the case of added specific analyser you should use contains instead of like  JCR-SQL operator
contains operator apply the configured analyser to searched keywords  (The indexed data is pre-treatment by the same analyser, so it's importent to have coherence between indexed data and  searched keyword) .
In the case of no configured analyser ( actual case) , We use the old logic : with like operator ( like use % option to have more results)

### CODE REVIEW

**Comment by thomasdelhomenie** [comment]
Please add a description to the PR to explain the changes.

**Comment by aboughzela** [comment]
@thomasdelhomenie  Descrption added to the PR

**Comment by thomasdelhomenie** [comment]
Thanks for the description @aboughzela .
Can't we use "contains" in both cases actually ?

Also, this requires a reindexation to be applied on existing data, right ?

**Comment by aboughzela** [comment]
>>Can't we use "contains" in both cases actually ?
With like operator , We have the possibility to use % option
so we use the contain only , if we have an custom analyse and if the keyword doesn't contain wildcard '*' 
exemple keyword : test*search ==> you should use like
 >>Also, this requires a reindexation to be applied on existing data, right ?
In the case of a user add a new analyser , you should re-index your data (this use case is not related for this patch)
This improvement doesn't add a new analyser but just treat this case

**Comment by thomasdelhomenie** [comment]
I am not sure to understand, my question is : is there a difference between "like %mykeyword%" and "contains mykeyword".

> This improvement doesn't add a new analyser but just treat this case

Right, the custom analyzer is not added in this PR. Actually why not ? I mean, it seems a good thing to have the IgnoreAccentAnalyzer analyzer by default (and of course let customers to change it if needed).

**Comment by aboughzela** [comment]
>>I am not sure to understand, my question is : is there a difference between "like %mykeyword%" and "contains mykeyword".
of course, no difference between "like %mykeyword%" and "contains mykeyword". 
but is possible with like to search my%keyword (by using my*Keyword) and not possible to have such possiblity without fuzzy search ( by defaut fuzzy is disabled for  PeopleSearchConnector and SpaceSearchConnector).
>>  Actually why not ?
yes , it's a good idea to configure analyser on social workspace (as collaboration)
but the old data need to be re-indexed.

**Comment by thomasdelhomenie** [comment]
> but is possible with like to search my%keyword (by using my*Keyword)

"contains my%keyword" does not work ?
Actually I find it quite strange to have different behavior depending on the analyzer. Will it still work if the customer choose another analyzer (not IgnoreAccentAnalyzer) ?

> yes , it's a good idea to configure analyser on social workspace (as collaboration)
> but the old data need to be re-indexed.

Yes, regarding the description issue (https://jira.exoplatform.org/browse/SOC-5283) it seems logic to apply this analyzer by default in the product. We must add a note in the upgrade notes to warn that if the customers want to benefit from this improvement for existing data they must re-index the social workspace (and we must add how to do it). I let you do the PR on https://github.com/exoplatform/platform-public-distributions/blob/develop/plf-packaging-resources/src/main/resources/RELEASE_NOTES.txt as well.
Also, what are the impacts if the customer do not reindex ?

**Comment by aboughzela** [comment]
>> Actually I find it quite strange to have different behavior depending on the analyzer. Will it still work if the customer choose another analyzer (not IgnoreAccentAnalyzer) ?

Yes , it still working with another analyser.
It'is easy to use just "contain" operator and add the analyser to social IgnoreAccentAnalyzer config.

>>what are the impacts if the customer do not reindex ?

the result is wrong if we try to search a word with Accent characters ( the property is indexed without  applying  analyser treatment and the search keyword use the anlyser)

**Comment by thomasdelhomenie** [comment]
> It'is easy to use just "contain" operator and add the analyser to social IgnoreAccentAnalyzer config.

Yes that looks like the good way to go.

> the result is wrong if we try to search a word with Accent characters ( the property is indexed without applying analyser treatment and the search keyword use the anlyser)

ok, so they will have the same behavior than before for the existing data and benefit from the improvement for the new data, they sounds acceptable.

**Comment by aboughzela** [comment]
@thomasdelhomenie 
No, we should re-index the old data to have a coherent index (index filed, without accent)
==>  that's the limit to add an analyser to old data

**Comment by thomasdelhomenie** [comment]
You mean that even for new data, they will not benefit from the improvement, but it will work like before ?

**Comment by aboughzela** [comment]
For the new data will  benefit from this improvement, but we can have  a search inconsistent with the old data (for exemple index data created on PLF 4.3.0, so we should re-index social data during upgrade)

**Comment by thomasdelhomenie** [comment]
ok, so it is acceptable IMO (I mean the behavior is acceptable even if the customer does not re-index). So I would go this way (add the Ignore accent analyzer by default), ok for you ?

**Comment by aboughzela** [comment]
Ok   I will update the PR as following :
* add the Ignore accent analyzer by default to social workspace
* use contain operator instead like

**Comment by thomasdelhomenie** [comment]
:+1: 
It means that the changes on JCR are not needed anymore, right ?

**Comment by aboughzela** [comment]
>> It means that the changes on JCR are not needed anymore, right ?

Not needed for this social PR, but I'll let the jcr improvement ( I have just added a method  to get the configured analyser from  searchIndex component ).

**Comment by thomasdelhomenie** [comment]
ok good, so I let you create a dedicated jira issue to track that.

**Comment by aboughzela** [comment]
@thomasdelhomenie 
Pull Request updated.

File: component/core/pom.xml [hunk 1]
```diff
@@ -309,6 +309,11 @@
       <scope>test</scope>
     </dependency>
     <dependency>
+      <groupId>org.exoplatform.ecms</groupId>
+      <artifactId>ecms-core-services</artifactId>
```
Comment (thomasdelhomenie) [inline component/core/pom.xml]: Why do we need this dependency ? We should avoid make social dependent on ecms if possible.

**Comment by thomasdelhomenie** [comment]
Where is added the default analyzer (IgnoreAccentAnalyzer) ? In another PR ?

**Comment by thomasdelhomenie** [comment]
Found it : https://github.com/exodev/commons/pull/40
Is it really necessary to have 2 Jira issues for these 2 PRs ?

**Comment by aboughzela** [comment]
>>IgnoreAccentAnalyzer : Is an old Analyser added to ECMS project (core/services module)

**Comment by aboughzela** [comment]
>>Why do we need this dependency ? We should avoid make social dependent on ecms if possible.

ECMS dependency added just for test porpose.

**Comment by thomasdelhomenie** [comment]
Is there any reason why this analyzer is in ECMS ?

**Comment by aboughzela** [comment]
>>Is there any reason why this analyzer is in ECMS ?

Simply because has been  created by ECMS Teams  (2010 related issue https://jira.exoplatform.org/browse/WCM-2860 )
I can move it to commons project ?

**Comment by thomasdelhomenie** [comment]
That's what I thought, unfortunately... :)
Moving it to commons would mean that existing installations that customized the workspaces definition would need to update their conf, so not teansparent. So I don't think worth it.

**Comment by aboughzela** [comment]
I think it possible by moving it in the same package as ecms
"org.exoplatform.services.jcr.analyzer.IgnoreAccentAnalyze"

**Comment by thomasdelhomenie** [comment]
Ah I did not see it was not in a ecms package. Why not so, if we do not change the package. Shouldn't it be directly in the JCR actually ?

**Comment by aboughzela** [comment]
It means that you think to move it from ECMS to JCR project in the same package ?

**Comment by thomasdelhomenie** [comment]
Yep. WDYT ?

**Comment by thomasdelhomenie** [comment]
Now that the analyzer has been moved to JCR, we can remove the ECMS dependency, right ?

**Comment by aboughzela** [comment]
Removed ECMS dependency : Done

File: pom.xml [hunk 2]
```diff
@@ -70,6 +70,7 @@
     <!-- Platform Project Dependencies -->
     <org.exoplatform.commons.version>4.4.x-SNAPSHOT</org.exoplatform.commons.version>
     <org.exoplatform.platform-ui.version>4.4.x-SNAPSHOT</org.exoplatform.platform-ui.version>
+
```
Comment (thomasdelhomenie) [inline pom.xml]: Useless change.

File: component/core/src/main/java/org/exoplatform/social/core/search/SpaceSearchConnector.java [hunk 3]
```diff
@@ -47,7 +47,7 @@ public SpaceSearchConnector(InitParams initParams, SpaceService spaceService) {
     SpaceFilter filter = new SpaceFilter();
     filter.setSpaceNameSearchCondition(query);
     filter.setSorting(sorting);
-    
+
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/search/SpaceSearchConnector.java]: Useless change.

**Comment by thomasdelhomenie** [comment]
:+1:

**Comment by aboughzela** [comment]
Remove Useless changes Done

**Comment by thomasdelhomenie** [comment]
Merged

**Pull request closed without merging**